csrc/host_ir/jit.cpp

```diff
@@ -681,74 +739,124 @@ class HostIrCompileDispatcher : public OptInDispatch {
     NVF_ERROR(
-        load_store_op->out()->isA<TensorView>(), "out must be a TensorView");
-    auto* in_tv = load_store_op->in()->as<Val>();
-    auto* out_tv = load_store_op->out()->as<Val>();
-    auto it = val_to_value_.find(in_tv);
-    NVF_ERROR(
-        it != val_to_value_.end(), "input tensor is not found in val_to_value");
-    llvm::Module* module = builder_.GetInsertBlock()->getParent()->getParent();
-    llvm::Value* in_tensor = it->second;
-    // Create a new tensor
-    llvm::Function* new_tensor_func = module->getFunction(kNewTensorFuncName);
-    llvm::Value* out_tensor =
-        builder_.CreateCall(new_tensor_func, {}, "out_tensor");
-
-    // Set the output tensor to the input tensor
-    llvm::Function* set_tensor_func = module->getFunction(kSetTensorFuncName);
-    builder_.CreateCall(set_tensor_func, {out_tensor, in_tensor});
+    load_store_op->out()->isA<TensorView>(), "out must be a TensorView");
+    auto* in_tv = load_store_op->in()->as<TensorView>();
+    auto* out_tv = load_store_op->out()->as<TensorView>();
+    llvm::Value* in_tensor = getOrCreateValue(in_tv, val_to_value_, builder_);
+    NVF_ERROR(in_tensor != nullptr, "in_tensor is nullptr");
+    // we assume all output tensors are already created, either through new or allocated
+    llvm::Value* out_tensor = getOrCreateValue(out_tv, val_to_value_, builder_);
 
-    // Bind the output tensor to val_to_value
+    llvm::Module* module = builder_.GetInsertBlock()->getParent()->getParent();
+    llvm::LLVMContext& context = builder_.getContext();
+  
+    if (out_tv->hasRoot()) {
+      std::optional<std::vector<int64_t>> permutation =
+          ir_utils::computePermutation(
+              out_tv->getRootDomain(), out_tv->getLogicalDomain());
+      NVF_ERROR(
+          permutation.has_value(),
+          "The logical domain of a Set.Permute is supposed to be a permutation"
+          " of the root domain: ",
+          out_tv);
+
+      llvm::Function* permute_out_func =
+          module->getFunction(kPermuteOutFuncName);
+      
+      // Create array of permutation values
+      llvm::ArrayType* perm_array_type = getInt64StaticArrayType(context, permutation.value().size());
+      llvm::Value* perm_array = builder_.CreateAlloca(perm_array_type, nullptr, "permutation");
+
+      for (size_t i = 0; i < permutation.value().size(); ++i) {
+        llvm::Value* gep = builder_.CreateInBoundsGEP(
+            perm_array_type, perm_array, 
+            {builder_.getInt32(0), builder_.getInt32(i)});
+        builder_.CreateStore(builder_.getInt64(permutation.value()[i]), gep);
+      }
+      
+      llvm::Type* int64_ptr_type = getInt64PtrType(context);
+      llvm::Value* perm_ptr = builder_.CreateBitCast(perm_array, int64_ptr_type);
+      llvm::Value* perm_size = builder_.getInt64(permutation.value().size());
+      if(out_tensor != nullptr) {
+        builder_.CreateCall(permute_out_func, {out_tensor, in_tensor, perm_ptr, perm_size});
```

Comment:
I'd remove this for YAGNI. In FusionExecutorCache, non-alias permutes go into kernels and won't appear as a top-level expr in the HostIrContainer.

Reply:
The same goes to kSetTensorOutFuncName.